examples/belief_propagation/bpexample.jl

```diff
@@ -28,41 +28,77 @@ function main()
   #Now do Simple Belief Propagation to Measure Sz on Site v
   nsites = 1
 
-  vertex_groups = nested_graph_leaf_vertices(
-    partition(partition(ψψ, group(v -> v[1], vertices(ψψ))); nvertices_per_partition=nsites)
+  Z = partition(ψψ, group(v -> v[1], vertices(ψψ)))
+  mts = message_tensors(ψψ, Z; contract_kwargs=(; alg="exact"))
+  mts = belief_propagation(ψψ, mts, 10; contract_kwargs=(; alg="exact"))
+  numerator_network = approx_network_region(
+    ψψ, mts, [(v, 1)]; verts_tn=ITensorNetwork([apply(op("Sz", s[v]), ψ[v])])
   )
-  mts = compute_message_tensors(ψψ; vertex_groups=vertex_groups)
-  sz_bp =
-    calculate_contraction(
-      ψψ, mts, [(v, 1)]; verts_tensors=ITensor[apply(op("Sz", s[v]), ψ[v])]
-    )[] / calculate_contraction(ψψ, mts, [(v, 1)])[]
+  denominator_network = approx_network_region(ψψ, mts, [(v, 1)])
+  sz_bp = contract(numerator_network)[] / contract(denominator_network)[]
 
   println(
     "Simple Belief Propagation Gives Sz on Site " * string(v) * " as " * string(sz_bp)
   )
 
   #Now do General Belief Propagation to Measure Sz on Site v
   nsites = 4
-  vertex_groups = nested_graph_leaf_vertices(
-    partition(partition(ψψ, group(v -> v[1], vertices(ψψ))); nvertices_per_partition=nsites)
+  Z = partition(
+    partition(ψψ, group(v -> v[1], vertices(ψψ))); nvertices_per_partition=nsites
   )
-  mts = compute_message_tensors(ψψ; vertex_groups=vertex_groups)
-  sz_bp =
-    calculate_contraction(
-      ψψ, mts, [(v, 1)]; verts_tensors=ITensor[apply(op("Sz", s[v]), ψ[v])]
-    )[] / calculate_contraction(ψψ, mts, [(v, 1)])[]
+  mts = message_tensors(ψψ, Z; contract_kwargs=(; alg="exact"))
```

Comment:
I think this could still be simplified a little bit (for example making a version of `message_tensors` that does the partitioning internally, and also a different interface for the keyword arguments since `contract_kwargs` is a bit confusing in this context), but I would be fine leaving that for a future PR.